parsing/extensions.ml

```diff
@@ -297,7 +297,36 @@ module Immutable_arrays = struct
 
   let of_pat expr = match expr.ppat_desc with
     | Ppat_array elts -> Iapat_immutable_array elts
-    | _ -> failwith "Malformed immutable array expression"
+    | _ -> failwith "Malformed immutable array pattern"
+end
+
+(** Strengthening *)
+module Strengthen = struct
+  type nonrec module_type =
+    { mty : Parsetree.module_type; mod_id : Longident.t Location.loc }
+
+  let extension_string = Language_extension.to_string Strengthen
+
+  (* Encoding: [S with M] becomes [functor (_ : S) -> M]. But remember
+     that this gets wrapped in an [[%extension.strengthen]] node, so
+     the what starts as [S with M] really becomes
+
+       {[
+         functor (_ : [%extension.strengthen]) ->
+           functor (_ : S) -> M
+       ]}
+  *)
+
+  let mty_of ~loc { mty; mod_id } =
+    (* See Note [Wrapping with make_extension] *)
+    Module_type.make_extension ~loc [extension_string] @@
+      Ast_helper.Mty.functor_ (Named (Location.mknoloc None, mty))
+        (Ast_helper.Mty.ident mod_id)
```

Comment:
This is a bit suspect, because it's stashing a *module* identifier where parsetree expects a *module type* identifier. All of the extensions packing and unpacking code is fine with this. But could a ppx take a pass looking for all module identifiers and then miss this one? I'm starting to doubt my choice of encoding around this point.

Reply:
What about `S with M` becoming

```
sig
  include S
  include module type of M
end
```

?  That seems like the closest analog to a pair we have, and at a quick glance `module type of` is the only way we have to reference the module language from the module type language without going via first-class modules.

Reply:
I'd replace `Pmty_ident` with `Pmty_alias` in the original encoding or go with @antalsz's suggestion (perhaps again, using `Pmty_alias` instead of `Pmty_typeof`).